Worley 2D

vladinator1000 opened the pull request:
Heya @johanhelsing , I love this crate! I'm working on a Worley implementation, it's almost there but I have some strange issues. This is the rendered output of this example:

```
$env:WGPU_BACKEND="dx12"; cargo run --example worley
```
I'm developing on windows and added the wgpu backend env var to reduce console warnings.

<img width="470" alt="worley_ERLTzJemo2" src="https://github.com/user-attachments/assets/b1e7fb05-43de-42cc-9f4c-847619234e50" />

I used this repo for inspiration: https://github.com/bevy-interstellar/wgsl_noise/blob/main/worley_vec2f.wgsl

Do you see where I'm making a mistake here? Looks like it doesn't handle negative coordinates well, because when  x and y are positive (top right) there are no visual issues.

vladinator1000 committed:
wip worley 2d

vladinator1000 committed:
replace fract with x - x.floor()

vladinator1000 committed:
use fract_gl to keep similarity with shader

vladinator1000 commented:
Turns out glam's `fract` implementation is different:  https://docs.rs/glam/latest/glam/f32/struct.Vec3.html#method.fract

I used `fract_gl` and it worked! Should be good for review now @johanhelsing. 

On the CPU:

<img width="960" alt="worley_czxM3hfQIA" src="https://github.com/user-attachments/assets/f78601fd-4cf0-4206-a10e-80cf57e177d1" />


The only thing that bothers me is that we use `value.y` [in the CPU demo](https://github.com/johanhelsing/noisy_bevy/pull/15/files#diff-bf36f76ccc5fe714319b6de0d26929f7ce538899748cfa5b97d52f5fa9de6dcaR31) it looks quite funky. Is it supposed to do that?

![image](https://github.com/user-attachments/assets/af3ae2d5-fd73-4a65-8c58-8d06100833ac)

The shader example looks pretty cool though:

<img width="960" alt="worley_shader_pKrcgZ6I5R" src="https://github.com/user-attachments/assets/89ca2c6e-cdce-4b66-a171-0863ab90c148" />

vladinator1000 committed:
add worley shader example

vladinator1000 committed:
add worley docs

vladinator1000 committed:
fix markdown warning

vladinator1000 committed:
update readme

vladinator1000 commented on README.md:
```diff
@@ -4,27 +4,26 @@
 [![docs.rs](https://img.shields.io/docsrs/noisy_bevy)](https://docs.rs/noisy_bevy)
 
-Simple stupid noise primitives for glam types (`Vec2`, `Vec3`) and wgsl.
+Simple stupid noise primitives for glam (`Vec2`, `Vec3`) and WGSL.
```
Tweaked the docs

vladinator1000 committed:
add new line

vladinator1000 commented on examples/README.md:
```diff
@@ -0,0 +1,19 @@
+# Examples
```
Example docs inspired by https://github.com/Razaekel/noise-rs/blob/develop/examples/README.md

vladinator1000 reviewed (commented):
Review notes

vladinator1000 committed:
follow mouse worley shader example

vladinator1000 commented:
Added mouse input to the example

https://github.com/user-attachments/assets/914b950a-9e3a-48d3-94f9-2dc91b201259

vladinator1000 committed:
fix worley

johanhelsing commented on assets/noisy_bevy.wgsl:
```diff
@@ -210,3 +210,69 @@ fn fbm_simplex_3d(pos: vec3<f32>, octaves: i32, lacunarity: f32, gain: f32) -> f
     return sum;
 }
+
+// Ported from https://github.com/bevy-interstellar/wgsl_noise
```
Should add a note that this is MIT-licensed?

johanhelsing commented on src/lib.rs:
```diff
@@ -297,6 +297,68 @@ pub fn fbm_simplex_3d(pos: Vec3, octaves: usize, lacunarity: f32, gain: f32) ->
+pub fn worley_noise_2d(pos: Vec2) -> Vec2 {
+    const K: f32 = 1.0 / 7.0;
+    const KO: f32 = 3.0 / 7.0;
+    let jitter = 1.0;
```
I'm not that famililar with worley noise. Would having jitter as an input be useful?

johanhelsing commented on src/lib.rs:
```diff
@@ -366,4 +428,21 @@ mod test {
         assert_debug_snapshot!(sample_3d_fn(|p| { fbm_simplex_3d(p, 5, 2.0, 0.5) }));
     }
+
+    // Enable for seeded version
+    // #[test]
+    // fn worley_2d_values_unchanged() {
```
Would be great to have a test, why is it disabled?

vladinator1000 replied:
It produces a different output every time, because it doesn't support seeding yet 😅

johanhelsing reviewed (commented):
Looks great, sorry for the slow response.

Did you check if it it produces similarish values on the CPU and GPU?

vladinator1000 committed:
fix worley noise, add test

vladinator1000 commented:
Thanks @johanhelsing, addressed your comments, but the CPU and GPU values look different and I'm not sure how to fix that.

CPU version `$env:WGPU_BACKEND="dx12"; cargo run --example worley`

<img width="960" alt="worley_SZKlDEfFbM" src="https://github.com/user-attachments/assets/eefaaf8f-ac30-4ad2-9fa2-dfc69ee72d19" />

GPU `$env:WGPU_BACKEND="dx12"; cargo run --example worley_shader`

<img width="960" alt="worley_shader_1Trhok36IR" src="https://github.com/user-attachments/assets/62601174-880f-4626-ab43-c61b1e9a1d3f" />

vladinator1000 committed:
ignore clippy::manual_swap lint

vladinator1000 commented on src/lib.rs:
```diff
@@ -297,6 +297,71 @@ pub fn fbm_simplex_3d(pos: Vec3, octaves: usize, lacunarity: f32, gain: f32) ->
 }
 
+/// Cellular noise
+#[allow(clippy::manual_swap)] // Allow to support no_std
```
Is this correct here or should I just use `std::mem::swap`?

vladinator1000 committed:
fix mem::swap lint

johanhelsing commented:
It's a bit unfortunate that we break the promise that gpu and cpu-noise are the same... It would be awesome if we could figure out what's causing the difference.

johanhelsing commented on README.md:
```diff
@@ -4,27 +4,26 @@
-![screenshot of an asteroid generated on the gpu and cpu](https://s3.johanhelsing.studio/dump/noisy_asteroid.png)
-
-## Implemented noise primitives:
+- `simplex_noise_2d`
+- `simplex_noise_2d_seeded`
+- `simplex_noise_3d`
+- `fbm_simplex_2d`
+- `fbm_simplex_2d_seeded`
+- `fbm_simplex_3d`
+- `worley_2d`
```
I think I prefer keeping the links `[]` here. As that makes the rust-doc links clickable

vladinator1000 replied:
Neat, I didn't know it could do that!

johanhelsing reviewed (commented):
I pretty closely at the two versions side-by-side, but couldn't find anything obvious causing the discrepancies :/ Guess maybe we'll have to go line-by-line and figure out what causes it

johanhelsing commented:
Turns out they're quite similar after all. I modified the examples quite heavily to make sure the inputs were the same, and they look similar.
![image](https://github.com/user-attachments/assets/b293b1dd-4d1d-49ae-89e1-92b0b615e873)

johanhelsing committed:
docs: Adjust example for cpu/gpu consistent results

johanhelsing commented:
Cleaned up the changes and made PR to your branch: https://github.com/vladinator1000/noisy_bevy/pull/1

Merge that and revert the changes that break the doc links, and this should be good to go :)

vladinator1000 committed:
Merge pull request #1 from johanhelsing/consistent-worley

docs: Adjust example for cpu/gpu consistent results

vladinator1000 committed:
update docs

vladinator1000 commented:
Wow, thanks @johanhelsing, turns out it was almost there ❤️ I fixed the docs.

johanhelsing reviewed (approved):
Thanks! This is a really cool addition. Thanks for sticking around even though my responses have been far in-between!
